docs: gitlab: add an explanation for required access token scope (#7952)
sourcegraph/sourcegraph: doc/admin/external_service/gitlab.md

  // ...
}
```

## Access token scopes

> Sourcegraph requires an access token with `api` permissions (and `sudo`, if you are using an `external` identity provider type). Below is an explanation for why we require this scope instead of e.g. just `read_user`, and what we may also use the access token for in the future.

Reviewer: ```suggestion
Sourcegraph requires an access token with `api` permissions (and `sudo`, if you are using an `external` identity provider type). These permissions are required for the following reasons:
```

> We are actively collaborating with GitLab on multiple fronts to improve our integration with them (e.g. through our [GitLab native integration](https://docs.gitlab.com/ee/integration/sourcegraph.html) and [working torwards better APIs Sourcegraph could use for querying repository permissions](https://gitlab.com/gitlab-org/gitlab/issues/20532)), so if you have any feedback please let us know!

Reviewer 2: ```suggestion
Our partnership with GitLab means we're continuously improving our [native code intelligence integration](https://docs.gitlab.com/ee/integration/sourcegraph.html) and [working towards better APIs Sourcegraph could use for querying repository permissions](https://gitlab.com/gitlab-org/gitlab/issues/20532). If you have any feedback, please let us know!
```

> We are actively collaborating with GitLab on multiple fronts to improve our integration with them (e.g. through our [GitLab native integration](https://docs.gitlab.com/ee/integration/sourcegraph.html) and [working torwards better APIs Sourcegraph could use for querying repository permissions](https://gitlab.com/gitlab-org/gitlab/issues/20532)), so if you have any feedback please let us know!

Reviewer: Move this section below the table.

Reviewer: ```suggestion
We are actively collaborating with GitLab to improve our integration (e.g. the [Sourcegraph GitLab native integration](https://docs.gitlab.com/ee/integration/sourcegraph.html) and [better APIs for querying repository permissions](https://gitlab.com/gitlab-org/gitlab/issues/20532)).
```

| [`GET /users`](https://docs.gitlab.com/ee/api/users.html#list-users) | `read_user` or `api` | If you are using an `external` identity provider type, used to discover user accounts. |
| [`GET /users/:id`](https://docs.gitlab.com/ee/api/users.html#single-user) | `read_user` or `api` | If using GitLab OAuth, used to fetch user metadata during the OAuth sign in process. |
| [`GET /projects/:id`](https://docs.gitlab.com/ee/api/projects.html#get-single-project) | `api` | (1) If using GitLab OAuth and repository permissions, used to determine if a user has access to a given _project_; (2) Used to query repository metadata (e.g. description) for display on Sourcegraph. |
| [`GET /projects/:id/repository/tree`](https://docs.gitlab.com/ee/api/repositories.html#list-repository-tree) | `api` | If using GitLab OAuth and repository permissions, used to verify a given user has access to the file contents of a repository within a project (i.e. does not merely have `Guest` permissions). |

Sourcegraph in the future may do more with the provided access token as well, including:

> - Allowing your Sourcegraph site-admins to perform large-scale code refactors, with Sourcegraph issuing and managing merge requests against your repositories on GitLab (only at your request as a Sourcegraph admin, of course).

Reviewer 2: ```suggestion
- Enabling Sourcegraph site-admins to perform large-scale code refactors, with Sourcegraph issuing and managing the merge requests on GitLab repositories, company-wide.
```
The "Enabling Sourcegraph site-admins to perform" section clarifies that this isn't an automated or system triggered event.

> - Allowing your Sourcegraph site-admins to perform large-scale code refactors, with Sourcegraph issuing and managing merge requests against your repositories on GitLab (only at your request as a Sourcegraph admin, of course).

Reviewer: The last 2 bullet points are redundant with what was stated earlier (are working with GitLab to improve their permissions API and we are improving our native integration), so delete them. Can you roll the first bullet point into the table? I'm not sure which GitLab endpoint we'll need for code refactors, but we just need convey that we will use the "write" part of the permissions to perform large-scale refactors in the future.

> - Using more efficient APIs to get repository and user permissions from GitLab more efficiently. We are actively working with GitLab to make this more efficient and use fewer requests, see https://gitlab.com/gitlab-org/gitlab/issues/20532

Reviewer 2: ```suggestion
- More efficient usage of GitLab APIs for fetching repository and user permissions, and we are [actively working with GitLab](https://gitlab.com/gitlab-org/gitlab/issues/20532) on these improvements.
```

> - Improving the GitLab native integration and Sourcegraph browser extension integration: https://docs.gitlab.com/ee/integration/sourcegraph.html

Reviewer 2: ```suggestion
- Improving the [GitLab native code intelligence integration](https://docs.gitlab.com/ee/integration/sourcegraph.html) and [Sourcegraph browser extension integration](../../integration/browser_extension.md).
```
